exporter/factory.go

```diff
@@ -0,0 +1,67 @@
+// Copyright 2019, OpenCensus Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package exporter
+
+import (
+	"github.com/spf13/viper"
+
+	"github.com/census-instrumentation/opencensus-service/processor"
+)
+
+// TraceDataExporter inherits functions from TraceDataProcessor, and additionally
+// adds some exporter-specific functions. This helps the collector core to
+// identify which TraceDataProcessors are Exporters and which are internal
+// processing components, so that better validation of pipelines can be done.
+type TraceDataExporter interface {
+	processor.TraceDataProcessor
+
+	// GetExportFormat gets the name of the format in which this exporter sends its data.
+	GetExportFormat() string
+}
+
+// TraceDataExporterFactory is an interface that builds a new TraceDataExporter based on
+// some viper.Viper configuration.
+type TraceDataExporterFactory interface {
+	// GetType gets the type of the TraceDataExporter created by this factory.
+	GetType() string
+	// NewFromViper takes a viper.Viper config and creates a new TraceDataExporter.
+	NewFromViper(cfg *viper.Viper) (TraceDataExporter, error)
+	// GetDefaultConfig returns the default configuration for TraceDataExporter
+	// created by this factory.
+	GetDefaultConfig() *viper.Viper
+}
+
+// MetricsDataExporter inherits functions from MetricsDataProcessor, and additionally
+// adds some exporter-specific functions. This helps the collector core to
+// identify which MetricsDataProcessors are Exporters and which are internal
+// processing components, so that better validation of pipelines can be done.
+type MetricsDataExporter interface {
```

Comment:
Why do we need this at all?

Reply:
Added it as a dual to the TraceDataExporter. I think that we should have the Metrics and Trace pipelines function in the same way.

Reply:
Happy to remove this though

Reply:
The same question applies to TraceDataExporter as well.

Reply:
Are you suggesting that the `Exporters`, since they implement `TraceDataProcessor`, just use that factory? If so, I think that is do-able, but I do kinda like having a separate factory for the separate type. Additionally, I think it will help when people are implementing exporters, since everything that they need to implement is then exporter focused.

Reply:
With the addition of `next`, this factory makes even more sense, since I think exporters should not have a `next` in their constructor

Reply:
I am suggesting that the FromViper returns directly a TraceDataProcessor.

Reply:
`TraceDataExporter` is a `TraceDataProcessor`, and I also think that we should make sure that `TraceDataExporterFactory` can only create `TraceDataExporter`s, incase someone tries to return just a `TraceDataProcessor`. We could get rid of `TraceDataExporter` altogether, although I kinda like having the distinction for the reasons in the comment